File: Python/reverseWords.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: reverseWords.py
@Function: 翻转单词顺序
@Link: https://leetcode-cn.com/problems/fan-zhuan-dan-ci-shun-xu-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-14
"""


class Solution:
    def reverseWords(self, s: str) -> str:
        s = s.strip()  # 删除首尾空格
        i = j = len(s) - 1
        res = []
        while i >= 0:
            while i >= 0 and s[i] != ' ':
                i -= 1  # 搜索首个空格
            res.append(s[i + 1: j + 1])  # 添加单词
            while s[i] == ' ':
                i -= 1  # 跳过单词间空格
            j = i  # j 指向下个单词的尾字符
        return ' '.join(res)  # 拼接并返回


# -----------------------------------
if __name__ == "__main__":
    str = "a good   example"

    solution = Solution()
    reverse_words = solution.reverseWords(str)
    print(f"The solution of this problem is : {reverse_words}")

File: C++/reverseWords.cpp
/**
 * @File    : reverseWords.cpp
 * @Brief   : 翻转单词顺序
 * @Link    : https://leetcode-cn.com/problems/fan-zhuan-dan-ci-shun-xu-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-14
*/

#include <iostream>
#include <string>

class Solution
{
public:
    std::string reverseWords(std::string s)
    {
        std::string res;
        int n = s.size();
        if (n == 0)
            return res;
        int right = n - 1;
        while (right >= 0)
        {
            //从后往前寻找第一字符
            while (right >= 0 && s[right] == ' ')
                right--;
            if (right < 0)
                break;

            //从后往前寻找第一个空格
            int left = right;
            while (left >= 0 && s[left] != ' ')
                left--;

            //添加单词到结果
            res += s.substr(left + 1, right - left);
            res += ' ';

            //继续往前分割单词
            right = left;
        }
        //去除最后一个字符空格
        if (!res.empty())
            res.pop_back();
        return res;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    std::string str = "the sky is blue";

    Solution solution;
    auto reverse_words = solution.reverseWords(str);
    std::cout << "The solution of this problem is : " << reverse_words << std::endl;

    return 0;
}
